From 6304838705e1b7408497ea2b25ebdcda05c83c7e Mon Sep 17 00:00:00 2001
From: "Ooi, Joyce" <joyce.ooi@intel.com>
Date: Mon, 2 Dec 2019 21:36:05 +0800
Subject: [PATCH 127/151] net: ethernet: altera: fix conversion of scaled_ppm
 to ppb

commit  242ef10b923b60bac73a937d93959c00d498d45f from
https://github.com/altera-opensource/linux-socfpga.git
branch is socfpga-5.4.64-lts

This patch fixes the conversion of scaled_ppm to ppb.

Signed-off-by: Ooi, Joyce <joyce.ooi@intel.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
Integrated-by: Jun Zhang <jun.zhang@windriver.com>
---
 drivers/net/ethernet/altera/intel_fpga_tod.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/altera/intel_fpga_tod.c b/drivers/net/ethernet/altera/intel_fpga_tod.c
index 0027f8b7c5e0..0d381818b3c7 100644
--- a/drivers/net/ethernet/altera/intel_fpga_tod.c
+++ b/drivers/net/ethernet/altera/intel_fpga_tod.c
@@ -98,7 +98,8 @@ static int intel_fpga_tod_adjust_fine(struct ptp_clock_info *ptp,
 	unsigned long flags;
 	unsigned long rate;
 	int ret = 0;
-	u64 ppb;
+	s64 ppb;
+	u64 new_ppb;
 
 	rate = clk_get_rate(priv->tod_clk);
 
@@ -107,9 +108,9 @@ static int intel_fpga_tod_adjust_fine(struct ptp_clock_info *ptp,
 	ppb *= 125;
 	ppb >>= 13;
 
-	ppb += NOMINAL_PPB;
+	new_ppb = (s32)ppb + NOMINAL_PPB;
 
-	tod_period = div_u64_rem(ppb << 16, rate, &tod_rem);
+	tod_period = div_u64_rem(new_ppb << 16, rate, &tod_rem);
 	if (tod_period > TOD_PERIOD_MAX) {
 		ret = -ERANGE;
 		goto out;
-- 
2.26.1

